pkg/storage/stores/tsdb/single_file_index.go

```diff
@@ -114,26 +116,60 @@ func (f *TSDBFile) Reader() (io.ReadSeeker, error) {
 // and translates the IndexReader to an Index implementation
 // It loads the file into memory and doesn't keep a file descriptor open
 type TSDBIndex struct {
-	reader      IndexReader
-	chunkFilter chunk.RequestChunkFilterer
+	reader         IndexReader
+	chunkFilter    chunk.RequestChunkFilterer
+	postingsClient PostingsClient
 }
 
 // Return the index as well as the underlying raw file reader which isn't exposed as an index
 // method but is helpful for building an io.reader for the index shipper
-func NewTSDBIndexFromFile(location string) (*TSDBIndex, GetRawFileReaderFunc, error) {
+func NewTSDBIndexFromFile(location string) (Index, GetRawFileReaderFunc, error) {
 	reader, err := index.NewFileReader(location)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	return NewTSDBIndex(reader), func() (io.ReadSeeker, error) {
+	tsdbIdx := NewTSDBIndex(reader, getPostingsClient(reader))
```

Comment:
I'm not sure whether passing a postingsClient to the TSDBIndex is the best approach.

```suggestion
	tsdbIdx := NewTSDBIndex(reader)
	tsdbIndex = NewCachedTSDBIndex(tsdbIndex)
```

I would have rather thought about inlineing the existing *TSDBIndex in a cached version, since `*TSDBIndex` is just an interface implementation of `IndexReader` and `RequestChunkFilterer`. Extend it with `PostingsClient`:

```go
type CachedTSDBIndex struct {
	*TSDBIndex
    PostingsClient
	cache Cache
}

func NewCachedTSDBIndex(tsdbIndex *TSDBIndex, backend Cache) *CachedTSDBIndex {
	return &CachedTSDBIndex{
		TSDBIndex: tsdbIndex,
		cache: backend,
	}
}
```

Then you can implement `ForPosting` from the `PostingsClient` interface.

I haven't completely thought it through, though.